File: python/src/blind150/twoPointers/BestTimeToSell.py
from typing import List
from src.utils.logging_util import get_logger

logger = get_logger(__name__)

class BestTimeToSell:
    def maxProfit(self, prices: List[int]) -> int:
        left, right, maxProfit = 0, 0, 0

        while right < len(prices):
            if prices[left] < prices[right]:
                profit = prices[right] - prices[left]
                maxProfit = max(profit, maxProfit)
            else:
                left = right
            right += 1

        logger.info(f"Max profit calculated: {maxProfit}")
        return maxProfit

if __name__ == "__main__":
    solution = BestTimeToSell()

    test_cases = [
        ([7, 1, 5, 3, 6, 4], 5),
        ([7, 6, 4, 3, 1], 0),
        ([1, 2, 3, 4, 5], 4),
        ([7, 6, 5, 4, 3], 0)
    ]

    for prices, expected in test_cases:
        logger.info(f"Checking prices: {prices}")
        result = solution.maxProfit(prices)
        logger.info(f"Max profit for {prices} is {result} (Expected: {expected})")

File: java/src/main/java/dsa/blind150/twoPointers/BestTimeToSell.java
package dsa.blind150.twoPointers;

import dsa.utils.LoggerUtil;
import org.slf4j.Logger;

public class BestTimeToSell {
    private static final Logger logger = LoggerUtil.getLogger();

    public int maxProfit(int[] prices) {
        int left = 0, right = 0, maxProfit = 0;

        while (right < prices.length) {
            if (prices[left] < prices[right]) {
                int profit = prices[right] - prices[left];
                maxProfit = Math.max(maxProfit, profit);
            } else {
                left = right;
            }
            right++;
        }
        logger.info("Max profit calculated: {}", maxProfit);
        return maxProfit;
    }

    public static void main(String[] args) {
        BestTimeToSell solution = new BestTimeToSell();

        int[] prices1 = {7, 1, 5, 3, 6, 4};
        int[] prices2 = {7, 6, 4, 3, 1};
        int[] prices3 = {5,1,5,6,7,1,10};

        logger.info("Max profit for prices1: {}", solution.maxProfit(prices1));
        logger.info("Max profit for prices2: {}", solution.maxProfit(prices2));
        logger.info("Max profit for prices3: {}", solution.maxProfit(prices3));
    }
}